src/mpb.cpp

```diff
@@ -41,35 +41,58 @@ typedef struct {
   double frequency;
   ndim dim;
   const fields *f;
+
+  /* for parallel efficiency, we first cache the epsinv data from each process, then
+     sum_to_all to synchonize, and then re-run set_maxwell_dielectric with the cached data */
+  double *cache; // array of 6*ncache eps_inv matrix entries in the order that they are needed
+  size_t ncache; // allocated size of the cache
+  size_t icache; // current position in the cache
+  bool use_cache; // whether we are using the cache
 } meep_mpb_eps_data;
 
 static void meep_mpb_eps(symmetric_matrix *eps, symmetric_matrix *eps_inv, const mpb_real r[3],
                          void *eps_data_) {
   meep_mpb_eps_data *eps_data = (meep_mpb_eps_data *)eps_data_;
-  const double *s = eps_data->s;
-  const double *o = eps_data->o;
-  double frequency = eps_data->frequency;
-  vec p(eps_data->dim == D3 ? vec(o[0] + r[0] * s[0], o[1] + r[1] * s[1], o[2] + r[2] * s[2])
-                            : (eps_data->dim == D2 ? vec(o[0] + r[0] * s[0], o[1] + r[1] * s[1]) :
-                                                   /* D1 */ vec(o[2] + r[2] * s[2])));
-  const fields *f = eps_data->f;
-
-  eps_inv->m00 = real(f->get_chi1inv(Ex, X, p, frequency));
-  eps_inv->m11 = real(f->get_chi1inv(Ey, Y, p, frequency));
-  eps_inv->m22 = real(f->get_chi1inv(Ez, Z, p, frequency));
-
-  ASSIGN_ESCALAR(eps_inv->m01, real(f->get_chi1inv(Ex, Y, p, frequency)), 0);
-  ASSIGN_ESCALAR(eps_inv->m02, real(f->get_chi1inv(Ex, Z, p, frequency)), 0);
-  ASSIGN_ESCALAR(eps_inv->m12, real(f->get_chi1inv(Ey, Z, p, frequency)), 0);
-  /*
-  master_printf("m11(%g,%g) = %g\n", p.x(), p.y(), eps_inv->m00);
-  master_printf("m22(%g,%g) = %g\n", p.x(), p.y(), eps_inv->m11);
-  master_printf("m33(%g,%g) = %g\n", p.x(), p.y(), eps_inv->m22);
-  master_printf("m12(%g,%g) = %g\n", p.x(), p.y(), eps_inv->m01);
-  master_printf("m13(%g,%g) = %g\n", p.x(), p.y(), eps_inv->m02);
-  master_printf("m23(%g,%g) = %g\n", p.x(), p.y(), eps_inv->m12);
-  */
-  maxwell_sym_matrix_invert(eps, eps_inv);
+  size_t i = eps_data->icache;
+
+  if (eps_data->use_cache) {
+    const double *cache = eps_data->cache;
+    eps_inv->m00 = cache[6*i];
+    eps_inv->m11 = cache[6*i+1];
+    eps_inv->m22 = cache[6*i+2];
+    ASSIGN_ESCALAR(eps_inv->m01, cache[6*i+3], 0);
+    ASSIGN_ESCALAR(eps_inv->m02, cache[6*i+4], 0);
+    ASSIGN_ESCALAR(eps_inv->m12, cache[6*i+5], 0);
+    maxwell_sym_matrix_invert(eps, eps_inv);
+  }
+  else {
+    // get cache pointer, doubling cache size as needed:
+    double *cache = i < eps_data->ncache ? eps_data->cache :
+        (eps_data->cache = (double*) realloc(eps_data->cache, 6 * (eps_data->ncache *= 2)));
```

Comment:
This line is missing a `sizeof(double)`: it should be `realloc(eps_data->cache, sizeof(double) * 6 * (eps_data->ncache *= 2)));`